File: neetcode150/09_heap_priority_queue/0355_design_twitter.py
"""
=====================================================================
LeetCode 355 : Design Twitter                                    (Medium)
https://leetcode.com/problems/design-twitter/
Category   : Heap / design
---------------------------------------------------------------------
PROBLEM
    postTweet(user, id), getNewsFeed(user) -> up to 10 most recent
    tweets of the user and everyone they follow; follow / unfollow.
---------------------------------------------------------------------
INTUITION
    Global increasing clock; each user's tweets form a chronologically
    sorted list. The feed is a k-way merge of these lists — a max-heap
    keyed by timestamp pops the newest tweet among all followed users.
---------------------------------------------------------------------
APPROACH 1 · Merge-k with a max-heap (best)
    Push the newest tweet of every followed user; pop the max; then
    push that user's next-newest; stop at 10. Heap stays ≤ #follows.
APPROACH 2 · Gather-all + sort (baseline)
    Collect every tweet of every followed user, sort by time desc.
---------------------------------------------------------------------
DEEP DIVE · k-way merge vs flat sort
    Flat: O(F·K log(F·K)) with F follows and K tweets each. Merge-k:
    O(F log F) per popped tweet — the classic external-sort pattern.
    Ties: we never pop more than one tweet of a user before its
    successor enters the heap, so ordering stays strictly chronological.
    The clock counts up per post, so "most recent" == "largest clock".
---------------------------------------------------------------------
TIME COMPLEXITY : getNewsFeed O(F log F); post/follow O(1).
MEMORY COMPLEXITY: O(T) for T tweets + follow edges.
=====================================================================
"""
import heapq, time, tracemalloc, collections


class Solution_1:
    """Merge-k feed with a max-heap of per-user tweet cursors."""

    def __init__(self):
        self.tweets = collections.defaultdict(list)   # user -> [(clock, id)]
        self.follows = collections.defaultdict(set)
        self.clock = 0

    def postTweet(self, userId, tweetId):
        self.tweets[userId].append((self.clock, tweetId))
        self.clock += 1

    def follow(self, followerId, followeeId):
        if followerId != followeeId:
            self.follows[followerId].add(followeeId)

    def unfollow(self, followerId, followeeId):
        self.follows[followerId].discard(followeeId)

    def getNewsFeed(self, userId):
        heap = []                                  # (-clock, id, author, idx)
        for author in {userId} | set(self.follows[userId]):
            tv = self.tweets[author]
            if tv:
                c, tid = tv[-1]
                heap.append((-c, tid, author, len(tv) - 1))
        heapq.heapify(heap)
        feed = []
        while heap and len(feed) < 10:
            neg_c, tid, author, idx = heapq.heappop(heap)
            feed.append(tid)
            if idx > 0:
                c, nid = self.tweets[author][idx - 1]
                heapq.heappush(heap, (-c, nid, author, idx - 1))
        return feed


class Solution_2:
    """Baseline: gather all tweets and sort."""

    def __init__(self):
        self.all = collections.defaultdict(list)
        self.follows = collections.defaultdict(set)
        self.clock = 0

    def postTweet(self, userId, tweetId):
        self.all[userId].append((self.clock, tweetId))
        self.clock += 1

    def follow(self, followerId, followeeId):
        if followerId != followeeId:
            self.follows[followerId].add(followeeId)

    def unfollow(self, followerId, followeeId):
        self.follows[followerId].discard(followeeId)

    def getNewsFeed(self, userId):
        pool = []
        for author in {userId} | set(self.follows[userId]):
            pool.extend(self.all[author])
        pool.sort()
        return [tid for _, tid in pool[-10:]][::-1]


if __name__ == "__main__":
    tracemalloc.start()
    t0 = time.perf_counter()
    s1, s2 = Solution_1(), Solution_2()
    ok = True


    def run(uid, want):
        global ok
        r1, r2 = s1.getNewsFeed(uid), s2.getNewsFeed(uid)
        ok &= r1 == want and r2 == want
        print(f"feed({uid}) -> {r1} want {want} {'PASS' if r1 == want and r2 == want else 'FAIL'}")


    s1.postTweet(1, 5); s2.postTweet(1, 5)
    run(1, [5])
    s1.follow(1, 2); s2.follow(1, 2)
    s1.postTweet(2, 6); s2.postTweet(2, 6)
    run(1, [6, 5])
    s1.unfollow(1, 2); s2.unfollow(1, 2)
    run(1, [5])
    s1.postTweet(2, 7); s2.postTweet(2, 7)
    run(1, [5])
    dt = time.perf_counter() - t0
    mem = tracemalloc.get_traced_memory()[1] / 1024
    tracemalloc.stop()
    print(f"{'PASS : all steps' if ok else 'FAIL'} | time: {dt * 1e6:.0f} us | mem: {mem:.0f} KB")

File: neetcode150/09_heap_priority_queue/0355_design_twitter.cpp
/*
 * =====================================================================
 * LeetCode 355 : Design Twitter                                    (Medium)
 * https://leetcode.com/problems/design-twitter/
 * Category   : Heap / Priority queue / design
 * ---------------------------------------------------------------------
 * PROBLEM
 *   Design a mini Twitter: postTweet(user, id), getNewsFeed(user)
 *   (10 most recent tweets from the user + everyone they follow,
 *   including themselves), follow(a, b), unfollow(a, b).
 * ---------------------------------------------------------------------
 * INTUITION
 *   Store every tweet with a global increasing timestamp. For a feed,
 *   merge the k most recent tweets from each followed user with a
 *   max-heap keyed by timestamp — like merging k sorted lists.
 * ---------------------------------------------------------------------
 * APPROACH 1 · Merge-k with max-heap (best)
 *   tweets: userId → list of (time, id), pushed chronologically.
 *   For getNewsFeed: push the LATEST tweet of every followed user,
 *   pop the newest, then push that user's next-latest. Stop at 10.
 * APPROACH 2 · Naive gather + sort (baseline)
 *   Collect all tweets of all followed users, sort by time desc, cut 10.
 * ---------------------------------------------------------------------
 * DEEP DIVE · Why the heap merge beats a flat sort
 *   Flat sort: O(f · k · log(f · k)) where f = follows, k = tweets each.
 *   Merge-k only ever holds f entries in the heap: O(f·log f) per
 *   popped tweet. For heavy accounts (k large) this is the classic
 *   k-way merge win. Timestamps: we just count up, so "recent" =
 *   "largest timestamp" — no wall-clock math needed.
 * ---------------------------------------------------------------------
 * TIME COMPLEXITY : getNewsFeed O(f · log f) with f followed users.
 * MEMORY COMPLEXITY: O(T) for T stored tweets + follow maps.
 * =====================================================================
 */
#include <iostream>
#include <vector>
#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <algorithm>
#include <chrono>
#include <sys/resource.h>

class Solution_1 {
    struct Tweet { int time; int id; };
    std::unordered_map<int, std::vector<Tweet>> tweets_;       // author → tweets
    std::unordered_map<int, std::unordered_set<int>> follows_; // user → follows
    int clock_ = 0;
public:
    void postTweet(int userId, int tweetId) {
        tweets_[userId].push_back({clock_++, tweetId});
    }
    void follow(int followerId, int followeeId) {
        if (followerId != followeeId) follows_[followerId].insert(followeeId);
    }
    void unfollow(int followerId, int followeeId) {
        follows_[followerId].erase(followeeId);
    }
    std::vector<int> getNewsFeed(int userId) {
        // newest-first heap: (time, tweetId, author, indexOfNext)
        using Node = std::tuple<int, int, int, int>;
        std::priority_queue<Node> heap;
        auto pushUser = [&](int author) {
            auto it = tweets_.find(author);
            if (it != tweets_.end() && !it->second.empty()) {
                size_t last = it->second.size() - 1;
                heap.emplace(it->second[last].time, it->second[last].id, author,
                             static_cast<int>(last));
            }
        };
        pushUser(userId);                                      // own tweets
        for (int f : follows_[userId]) pushUser(f);            // followed users
        std::vector<int> feed;
        while (!heap.empty() && feed.size() < 10) {
            auto [time, id, author, idx] = heap.top();
            heap.pop();
            feed.push_back(id);
            if (idx > 0) {                                     // next-latest tweet
                const auto& tv = tweets_[author][idx - 1];
                heap.emplace(tv.time, tv.id, author, idx - 1);
            }
        }
        return feed;
    }
};

class Solution_2 {
    std::unordered_map<int, std::vector<std::pair<int,int>>> all_; // user → (time,id)
    std::unordered_map<int, std::unordered_set<int>> follows_;
    int clock_ = 0;
public:
    void postTweet(int userId, int tweetId) {
        all_[userId].emplace_back(clock_++, tweetId);
    }
    void follow(int followerId, int followeeId) {
        if (followerId != followeeId) follows_[followerId].insert(followeeId);
    }
    void unfollow(int followerId, int followeeId) {
        follows_[followerId].erase(followeeId);
    }
    std::vector<int> getNewsFeed(int userId) {
        std::vector<std::pair<int,int>> pool;
        auto add = [&](int author) { for (auto& p : all_[author]) pool.push_back(p); };
        add(userId);
        for (int f : follows_[userId]) add(f);
        std::sort(pool.begin(), pool.end());                   // ascending time
        std::vector<int> feed;
        for (size_t i = pool.size(); i-- > 0 && feed.size() < 10;)
            feed.push_back(pool[i].second);                    // newest first
        return feed;
    }
};

static long long nowUs() {
    return std::chrono::duration_cast<std::chrono::microseconds>(
        std::chrono::steady_clock::now().time_since_epoch()).count();
}
static long memKb() { struct rusage r{}; getrusage(RUSAGE_SELF, &r); return r.ru_maxrss; }

int main() {
    const long mem0 = memKb();
    auto t0 = nowUs();
    Solution_1 s1;
    Solution_2 s2;
    bool allOk = true;
    auto run = [&](int uid, const std::vector<int>& expect) {
        auto r1 = s1.getNewsFeed(uid);
        auto r2 = s2.getNewsFeed(uid);
        bool ok = r1 == expect && r2 == expect;
        allOk = allOk && ok;
        std::cout << "feed(" << uid << ") → [" ;
        for (size_t i = 0; i < r1.size(); ++i) std::cout << (i ? "," : "") << r1[i];
        std::cout << "] A2=[" ;
        for (size_t i = 0; i < r2.size(); ++i) std::cout << (i ? "," : "") << r2[i];
        std::cout << "] want=[" ;
        for (size_t i = 0; i < expect.size(); ++i) std::cout << (i ? "," : "") << expect[i];
        std::cout << "] " << (ok ? "PASS" : "FAIL") << "\n";
    };
    s1.postTweet(1, 5); s2.postTweet(1, 5);
    run(1, {5});
    s1.follow(1, 2); s2.follow(1, 2);
    s1.postTweet(2, 6); s2.postTweet(2, 6);
    run(1, {6, 5});
    s1.unfollow(1, 2); s2.unfollow(1, 2);
    run(1, {5});
    s1.postTweet(2, 7); s2.postTweet(2, 7);
    run(1, {5});
    auto t1 = nowUs();
    std::cout << (allOk ? "PASS : all steps" : "FAIL") << " | time: " << (t1 - t0)
              << " us | mem: " << (memKb() - mem0) << " KB\n";
    return 0;
}
